File: android/src/main/java/com/iconshot/detonator/element/IconElement.java
package com.iconshot.detonator.element;

import android.graphics.Typeface;
import android.util.TypedValue;
import android.widget.TextView;

import com.iconshot.detonator.Detonator;
import com.iconshot.detonator.helpers.ColorHelper;
import com.iconshot.detonator.helpers.CompareHelper;
import com.iconshot.detonator.helpers.ContextHelper;
import com.iconshot.detonator.helpers.IconHelper;
import com.iconshot.detonator.helpers.PixelHelper;

public class IconElement extends Element<TextView, IconElement.Attributes> {
    private int defaultColor;

    public IconElement(Detonator detonator) {
        super(detonator);
    }

    @Override
    public Class<Attributes> getAttributesClass() {
        return Attributes.class;
    }

    @Override
    protected TextView createView() {
        TextView view = new TextView(ContextHelper.context);

        defaultColor = view.getCurrentTextColor();

        return view;
    }

    @Override
    protected void patchView() {
        String name = attributes.name;
        String currentName = currentAttributes != null ? currentAttributes.name : null;

        boolean patchName = forcePatch || !CompareHelper.compareObjects(name, currentName);

        if (patchName) {
            String font = IconHelper.getFont(name);

            if (font != null) {
                Typeface typeface = IconHelper.getTypeface(font);

                view.setTypeface(typeface);

                Character iconChar = IconHelper.getIcon(name);

                view.setText(iconChar != null ? String.valueOf(iconChar) : "");
            } else {
                view.setTypeface(null);
                view.setText("");
            }
        }
    }

    protected void patchFontSize(Float fontSize) {
        float value = ((Integer) PixelHelper.dpToPx(fontSize != null ? fontSize : 16)).floatValue();

        view.setTextSize(TypedValue.COMPLEX_UNIT_PX, value);
    }

    protected void patchColor(Object color) {
        Integer tmpColor = ColorHelper.parseColor(color);

        view.setTextColor(tmpColor != null ? tmpColor : defaultColor);
    }

    static class Attributes extends Element.Attributes {
        String name;
    }
}

File: ios/Detonator/Sources/Detonator/Element/IconElement.swift
import UIKit

class IconElement: Element {
    override func decodeAttributes() -> IconAttributes? {
        return super.decodeAttributes()
    }
    
    override func createView() -> TextView {
        return TextView()
    }
    
    override func setUpView() -> Void {
        let view = view as! TextView
        
        view.isUserInteractionEnabled = true
    }

    override func patchView() -> Void {
        let view = view as! TextView
        
        let attributes = attributes as! IconAttributes
        let prevAttributes = prevAttributes as! IconAttributes?
        
        let name = attributes.name
        let prevName = prevAttributes?.name
        
        let patchNameBool = forcePatch || name != prevName
        
        if patchNameBool {
            let font = IconHelper.getFont(name: name)
            
            if let font = font {
                view.font = UIFont(name: font, size: view.font.pointSize)
                
                view.text = IconHelper.getIcon(key: name)
            } else {
                view.font = UIFont.systemFont(ofSize: view.font.pointSize)
                
                view.text = nil
            }
        }
    }
    
    override func patchFontSize(fontSize: Float?) -> Void {
        let view = view as! TextView
        
        let value = CGFloat(fontSize ?? 16)
        
        view.font = view.font.withSize(value)
    }
    
    override func patchColor(color: StyleColor?) -> Void {
        let view = view as! TextView
        
        view.textColor = color != nil ? ColorHelper.parseColor(color: color!) : nil
    }
    
    class IconAttributes: Attributes {
        var name: String

        required init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            
            name = try container.decode(String.self, forKey: .name)
            
            try super.init(from: decoder)
        }
        
        private enum CodingKeys: String, CodingKey {
            case name
        }
    }
}
